modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py

```diff
@@ -0,0 +1,172 @@
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsampling_strategy import (
+    get_tensors_subset,
+)
+from modyn.trainer_server.internal.trainer.remote_downsamplers.remote_rs2_downsampling import RemoteRS2Downsampling
+
+
+def test_init():
+    pipeline_id = 0
+    trigger_id = 0
+    batch_size = 32
+    params_from_selector = {"replacement": True, "downsampling_ratio": 50}
+    per_sample_loss = None
+    device = "cpu"
+
+    downsampler = RemoteRS2Downsampling(
+        pipeline_id, trigger_id, batch_size, params_from_selector, per_sample_loss, device
+    )
+
+    assert downsampler.pipeline_id == pipeline_id
+    assert downsampler.trigger_id == trigger_id
+    assert downsampler.batch_size == batch_size
+    assert downsampler.device == device
+    assert not downsampler.forward_required
+    assert not downsampler.supports_bts
+    assert downsampler._all_sample_ids == []
+    assert downsampler._subsets == []
+    assert downsampler._current_subset == -1
+    assert downsampler._with_replacement == params_from_selector["replacement"]
+    assert downsampler._max_subset == -1
+    assert downsampler._first_epoch
+
+
+def test_inform_samples():
+    pipeline_id = 0
+    trigger_id = 0
+    batch_size = 32
+    params_from_selector = {"replacement": True, "downsampling_ratio": 50}
+    per_sample_loss = None
+    device = "cpu"
+
+    downsampler = RemoteRS2Downsampling(
+        pipeline_id, trigger_id, batch_size, params_from_selector, per_sample_loss, device
+    )
+
+    sample_ids = [1, 2, 3, 4, 5]
+    forward_output = torch.randn(5, 10)
+    target = torch.randint(0, 10, (5,))
+
+    downsampler.inform_samples(sample_ids, forward_output, target)
+
+    assert downsampler._all_sample_ids == sample_ids
+    downsampler.inform_samples(sample_ids, forward_output, target)
+    assert downsampler._all_sample_ids == 2 * sample_ids
+    # Now it should not change anymore
+    downsampler.select_points()
+    downsampler.inform_samples(sample_ids, forward_output, target)
+    assert set(downsampler._all_sample_ids) == set(sample_ids)
+    assert len(downsampler._all_sample_ids) == 2 * len(sample_ids)
+
+
+def test_multiple_epochs_with_replacement():
+    pipeline_id = 0
+    trigger_id = 0
+    batch_size = 32
+    params_from_selector = {"replacement": True, "downsampling_ratio": 50}
+    per_sample_loss = None
+    device = "cpu"
+
+    downsampler = RemoteRS2Downsampling(
+        pipeline_id, trigger_id, batch_size, params_from_selector, per_sample_loss, device
+    )
+    with torch.inference_mode(mode=(not downsampler.requires_grad)):
+        sample_ids = list(range(10))
+        data = torch.randn(10, 10)
+        target = torch.randint(0, 10, (10,))
+
+        for _ in range(3):
+            downsampler.inform_samples(sample_ids, data, target)
```

Comment:
Each call to `inform_samples` should be provided with a different set of `sample_ids`

Reply:
Yeah it does not make a difference here, as we just test the shape. But naturally they should be different because,

In `sample_and_batch`. In the `pytorch_trainer.py`, we first iterate over the dataloader and keep informing each batch in `_iterate_dataloader_and_compute_scores` https://github.com/eth-easl/modyn/blob/ac66eafc55c9089f938436e6255062df805bde9a/modyn/trainer_server/internal/trainer/pytorch_trainer.py#L836

the `sample_ids` come from the dataloader and should be naturally distinct right? (they are keys of the samples)

Reply:
No, they are not. What differs is the model output (on which true downsamplers sample), but the list of samples is always the same, since the trigger training set from the selector does not change between epochs. Since RS2 only relies on the IDs, it should not matter. The IDs will in all cases be identical across epochs.

Reply:
I think we have a mistake. I am saying the consecutive calls to `inform_samples` within two `select_points` calls should contain different sample ids.

I copy the code of `_iterate_dataloader_and_compute_scores` here:
```
        for batch_number, batch in enumerate(dataloader):
            self.update_queue(AvailableQueues.DOWNSAMPLING, batch_number, number_of_samples, training_active=False)

            sample_ids, target, data = self.preprocess_batch(batch)
            number_of_samples += len(sample_ids)

            with torch.inference_mode(mode=(not self._downsampler.requires_grad)):
                with torch.autocast(self._device_type, enabled=self._amp):
                    # compute the scores and accumulate them
                    model_output = self._model.model(data)
                    embeddings = self.get_embeddings_if_recorded()
                    self._downsampler.inform_samples(sample_ids, model_output, target, embeddings)
```

You see. We load batch from the dataloader. The first batch must have different sample ids than the second batch's sample ids.

I am not talking about sample ids across epochs. Those definitely do not change.

Reply:
i.e. if we have
```
downsampler.inform_samples(...)
downsampler.select_points(...)
downsampler.inform_samples(...)
```
Then the first inform_samples call can have the same sample ids as the second inform_samples.

But when we do 

```
downsampler.inform_samples(...)
downsampler.inform_samples(...)
downsampler.select_points(...)
downsampler.inform_samples(...)
downsampler.inform_samples(...)
```
Suppose the whole dataset contains two batches. Then the first two `inform_samples` calls should contain different `sample_ids`.

Reply:
Anyway, I think it does not really make a difference here to use different sample ids. But I still do think consecutive `inform_samples` calls (without `select_points` call in the middle) should contain distinct sample ids.

Reply:
I understand your point and agree with your description, but I still don't understand why you are suggesting it here :D The code is this

```
    with torch.inference_mode(mode=(not downsampler.requires_grad)):
        sample_ids = list(range(10))
        data = torch.randn(10, 10)
        target = torch.randint(0, 10, (10,))

        for _ in range(3):
            downsampler.inform_samples(sample_ids, data, target)
            selected_ids, weights = downsampler.select_points()
```

so the loop is the epoch loop (!). Since `sample_ids = list(range(10))` we don't have duplicate samples in the same epoch and consistent samples across epochs. This is exactly like you describe. I am not sure if I am missing something or you just confused this loop with something else. I am merging this for now and happy to do a follow up PR in case I am missing something here

Reply:
" this unit test, we only keep calling inform_samples(...) without calling select_points(...),"

i dont get it. isn't it directly below :D?